### ISSUE TITLE
CardView: Dashboard suggestions?

### ISSUE THREAD

**Issue by cdhorn** [cdhorn/CardView#196]
I would like suggestions for what people think a "dashboard" card view page should provide.

I ask because it seems like there would be a lot of overlap with existing gramplets.

**Comment by fxtmtrsine** [comment]
I love that you made the Dashboard Card show Stat's on the current family tree 🥇 
![2022-06-14 16_15_42-Example family tree - Tag - Gramps](https://user-images.githubusercontent.com/32380545/173743883-87251385-4869-40c8-acc7-3c22c032ad49.png)

Would it make sense to always show the section title

![2022-06-15 15_16_17-main - Tag - Gramps](https://user-images.githubusercontent.com/32380545/173743903-23f4d725-b9e7-4f8c-aa49-d3a7cb9df432.png)

otherwise you get this

![2022-06-15 15_16_32-main - Tag - Gramps](https://user-images.githubusercontent.com/32380545/173743898-be032491-c023-42f6-a83b-47e1e097374c.png)

maybe a toggle to show/hide all ?

**Comment by fxtmtrsine** [comment]
Noticed that **Pin copy of current view** does not work, maybe a work in progress?

**Comment by fxtmtrsine** [comment]
As requested some ideas based on more stats that I'd like to see, hopefully others and yourself agree?

* Individual Stats  - show the name of the Home Person and Active Person ( not sure it would be useful for the other objects?)
* Information on Tags and Bookmarks?
* This probably applies to all Card views but when used on a smaller screen I find myself wishing that each of the other columns adjusted  and appeared below the others columns so I don't have to scroll to the right and only have to keep scrolling down the page?
* Information on how many groups of people are linked to each other in the family tree, but not linked to people in other groups ( I keep multiple unrelated familytrees in Gramps and use my bookmarks to jump between them and tags to differentiate!  ) Not sure If I'm clear on what I mean here? example: Group 1 - John Smith[ID5600], 800 people, 6 generations  ..... Total 23 Groups, 14,300 people ( should match total people in family tree )
* Not sure what it's called but stats for the Home Person only that by generation how complete the ancestor tree is eg:  

gen 1 - Home Persons name
gen 2 - Parents 2 of 2
gen 3 - grandparents 3 of 4
gen 4 - great grandparents 5 of 8
etc...
Total complete 10 of 14 ( percentage found )

**Comment by cdhorn** [comment]
Thanks for the suggestions! This view is not done, it's kind of sandbox/work in progress.  It is heavy in sense it parses the full tree like the last changed gramplet so is slow to load on a big tree and you don't have the yield trick in play to help the ui.  I am looking at / pondering that. The current groups grew out of Doug Blank's stats gramplet.

> This probably applies to all Card views but when used on a smaller screen I find myself wishing that each of the other columns adjusted and appeared below the others columns 

In the template editor, under layout tab is the layout for each object/page view the card view renders.  Append next group controls if next group in the list is appended to bottom of current one or starts a new column as well as which groups to make visible and which not to.  You can enable scrolling for the groups as well, or render them in tabbed notebook. I have not added the tab for the dashboard view layout yet but there will be one.

Since you can have multiple templates you can have different layouts, and then ctrl-left click to get template context menu to quickly change the active one.

**Comment by emyoulation** [comment]
> Individual Stats - show the name of the Home Person and Active Person ( not sure it would be useful for the other objects?)

How about a People Filter option on Stats? There are Add-on Rules for Home Person Active Person. And for Degrees of Separation. (If the filter becomes too complex or multi-stage, you can run it in the Add/Remove Tag Tool and apply a new Tag the Filter results. Then use a faster Tag filter to speed up operation.)

But once applying filters becomes a possibility, perhaps it makes more sense to have a copy of the Stats view mode in the People category. So the Dashboard category Stats does the Whole Tree, the People category respects the Filter Gramplet & Search bar to limit the scope.

For Privacy, finding 'Living' would be helpful. 

Haven't downloaded yet, it's there a Event Roles stat panel?

**Comment by cdhorn** [comment]
> In the template editor, under layout tab is the layout for each object/page view the card view renders. Append next group controls if next group in the list is appended to bottom of current one or starts a new column as well as which groups to make visible and which not to. You can enable scrolling for the groups as well, or render them in tabbed notebook. I have not added the tab for the dashboard view layout yet but there will be one.

@fxtmtrsine I should have also mentioned in the layout editor you can drag and drop to rearrage the group order as well.

**Comment by cdhorn** [comment]
> How about a People Filter option on Stats?

I'd need to look closer but you can't add a filter gramplet to the Dashboard view because there is no anchoring navigation object type for the dashboard.

The cardviews for the other objects are really for a single object so I don't think filters really can apply to them, at least not with this design.  Some sort of stats group may make sense. I really ought to find a way to make the groups into plugins so other devs can contribute stuff if they want like I did with the calculated fields and status indicators.

> For Privacy, finding 'Living' would be helpful.

Agree. Note at moment I just provide counts, it's not like Doug's original that pulls up quick view reports for things.

> is there an Event Roles stat panel?

Oh yeah, I can add one for event roles, associations and maybe some other stuff as well.

**Comment by cdhorn** [comment]
@fxtmtrsine pinned copies work, layout editor works for dashboard view now and the apply bug fixed, notes, tags and bookmarks panels added.

**Comment by cdhorn** [comment]
> For Privacy, finding 'Living' would be helpful.

This done too now and I also have count for "living and not marked private"

What I wanted to try to do with this was to try to show counts for things that might give a birds eye view of potential areas in a tree with data quality and documentation issues. 

Tags and bookmarks don't really fit that profile, and I'm unsure if groups breaking out role counts really do either. But then anyone can enable/disable what they want so I guess that is okay.

**Comment by PQYPLZXHGF** [comment]
> What I wanted to try to do with this was to try to show counts for things that might give a birds eye view of potential areas in a tree with data quality and documentation issues.

I like it, thank you make the card into a real dashboard :)

A few more nice to have things to show optionally:
* [Researcher](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Settings#Researcher)/[Database Owner Information](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Tools#Edit_Database_Owner_Information)
* For Quality Statistics add the information from the [Verify the data tool](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Tools#Verify_the_Data) 
* Info from [What's Next](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Gramplets#What.27s_Next) gramplet
* A section to show info from the [Records Gramplet](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Gramplets#Records)
* A section to show info from the [Session Log Gramplet](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Gramplets#Session_Log)  ( A big bonus if it saves this information to a external file between sessions 😜 )

**Comment by PQYPLZXHGF** [comment]
> Not sure what it's called but stats for the Home Person only that by generation how complete the ancestor tree is eg:
> gen 1 - Home Persons name

@fxtmtrsine  Sound like you are describing something simlar to the little report on the bottom of the [Pedigree Gramplet](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Gramplets#Pedigree) that provides a "Breakdown by Generation" 

![Pedigree-Gramplet-detached-50](https://user-images.githubusercontent.com/212355/174039755-16577754-17df-4441-a92c-2778bb06c065.png)

**Comment by emyoulation** [comment]
In the future, you might want to experiment with graphing progress of some or all of these stats.
You might want to add a (optional?) logging function to collect data points. 

![download.jpeg](https://user-images.githubusercontent.com/69127217/174088092-276468b1-487e-446c-9055-44269f71bbfd.jpeg)

It would probably need to store tuples since you will have an evolving set of data element. Which means that any graph/analytics would need to be tolerant of missing/extra datapoint groupings.

**Comment by emyoulation** [comment]
In the future, you might want to experiment with graphing progress of some (or all) of these stats.
You might want to add a (optional?) logging function to collect data points. 

![download.jpeg](https://user-images.githubusercontent.com/69127217/174088092-276468b1-487e-446c-9055-44269f71bbfd.jpeg)

It would probably need to store tuples since you will have an evolving set of data element. Which means that any graph/analytics would need to be tolerant of missing/extra datapoint groupings.

**Comment by cdhorn** [comment]
> In the future, you might want to experiment with graphing progress of some (or all) of these stats.

That is an interesting idea, capturing certain metrics while a tree is being developed for possible analysis. I'm not aware of any other genealogy software doing that.

Question is what makes sense to track and what does not and where do you store the data?  Over time it could grow to a sizeable amount, and ideally it belongs in the database and is accounted for as part of the data model.

Value wise not sure how much it delivers in the end though, time is probably better spent on features and things we do not have at all today or that are not fully developed that could be more useful to a wider audience.

**Comment by emyoulation** [comment]
GitHub is doing some weird stuff in the comment function via my Android phone. Instead of a blank new comment, if if inserting old content along with my new comment. The last one had invisibly prefixes a previous message until the message was posted. Had to edit to prune that out.

Sorry.

**Comment by patsyblefebre** [comment]
Love the addon, really appreciate the effort that has gone into it Thank you 🙌✨🎶

> @fxtmtrsine Individual Stats - **show the name of the Home Person and Active Person** ( not sure it would be useful for the other objects?)

> @PQYPLZXHGF [Researcher](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Settings#Researcher)/[Database Owner Information](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Tools#Edit_Database_Owner_Information)


Instead of placing these in the Individual stats section maybe create a "Family Tree Summary/Details" section (at the top of the page like on other cards) 

![cardview-view-person-highlight top section](https://user-images.githubusercontent.com/107233751/174201235-f6aa25c5-5dbf-4dc3-a2c6-a1f6c082b02f.png)


that has some of the details from the Family Tree manager  

![](https://user-images.githubusercontent.com/107233751/174200237-90e556a3-67e4-4b59-ae0c-d4d0b32d8b05.png)

Like the; Family Tree name  /  Database  Type / Researcher info / (and then)  **show the name of the Home Person and Active Person** ?

Hopefully sometime in the future Gramps will allow you to add a summary description to the family tree name entry like the [Isotammi  XML export](https://gramps-project.org/wiki/index.php/Addon:Isotammi_addons#Isotammi_XML_export) allows you to. 

![Isotammi-XML-export-Isotammi-51](https://user-images.githubusercontent.com/107233751/174200978-90b557ff-e83a-45bc-8c51-fa97c88da7a8.png)

**Comment by cdhorn** [comment]
Oh of course!  Excellent idea with the description!

**Comment by cdhorn** [comment]
Going forward the database description will be stored in a note with custom type "Database Description"

Should add it to the official note types really, but then I still haven't figured out why my "Proof Argument" note type blows up the tests.

**Comment by fxtmtrsine** [comment]
> Going forward the database description will be stored in a note with custom type "Database Description"

What happens if two or more notes of custom type "Database Description"?

Note my suggestions are just that , glad it helps define what it is becoming.

Thanks for the fixes, 


@patsyblefebre > Instead of placing these in the Individual stats section maybe create a "Family Tree Summary/Details" section (at the top of the page like on other cards)

I like that, maybe an image or background banner could be shown  using a custom media type "Database Logo"?

**Comment by cdhorn** [comment]
A couple issues to sort but the header will sort of look like this if any thoughts on anything else I should try to add to it.
![Screenshot from 2022-06-17 00-13-15](https://user-images.githubusercontent.com/43528418/174223728-10a2f62b-eb0d-48bb-a81b-21bada6ed385.png)
I thought about using the person cards for the home and active people but they consume more space and I think this view should focus more on the database.

**Comment by daleathan** [comment]
Looks good, nice and simple.

> ...if any thoughts on anything else ...

Instead of "**Database Information**" name it "**Family Tree Information**" or "...Summary" in keeping with the rest of Gramps

> [Family Trees are what Gramps calls the database structure used to store and organize genealogical data.](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Manage_Family_Trees#Family_Trees_manager_window)

**Comment by cdhorn** [comment]
Thank you!  Family Tree Summary it is, fits better with what the view is trying to do.

**Comment by emyoulation** [comment]
Suggest that you set the MantisBT to monitor [Gramps Feature Request 11817](https://gramps-project.org/bugs/view.php?id=11817). And add a note in it's comments that remind you to revisit the CardView Dashboard note when/if that feature is implemented.

There are 2 Note based functionalities requested: 
1) define a ReadMe note that is inserted as plain text at the top of Gramps archives (backups). Those ReadMe files describe the archive & possibly has a Report with stats.
2) a Introductory Note. This Note overrides the default content of the Welcome Gramplet. It would allow the research to have a "Guided tour" or preface.

You will have to decide if you could leverage such note for your dashboard. And joining the MantisBT discussion would let you have input on the implementation to avoid double work.

**Comment by emyoulation** [comment]
A note within the XML serves a different purpose. (Although that might be integrated with the Introductory note.)

The ReadMe.txt is outside the not-generally-known .Gramps file extension XML file inside the .gpkg zipped archive.

You don't need an XML file reader to parse it. You don't need to search through multiple megabytes looking for the Note chunk... nor drill down to find a Note of a specific type... nor decipher the markup within that note.

The ReadMe is at the top level of the archive along with 1 .Gramps file & a Media folder.

If you've explored the .gpkg files, you'll have found that they are recursively compressed too! The .gramps file withing the .gpkg is ZIP compressed and has to be expanded before it can be read into an XML reader. And since the file inside the compressed .gramps file is an uncompressed .gramps file with AN IDENTICAL NAME, it adds a layer of file management nightmare.

One last point, .gpkg files are not a file type that will be accepted as an attachment by most eMail spam filters or as an uploadable filetype. If Gramps allowed us to import a ZIP archive (after verifying it confirmed to what we currently expect in a .gpkg or a compressed .gramps file, we could just change the filetype & upload. The person importing would not be able to double-click to launch Gramps & automatically sub-launch the import. But they'd be able to use the import from within Gramps.

Basically, this whole section of Gramps needs streamlining because of all of the excessive complexity which adds no value.

**Comment by PQYPLZXHGF** [comment]
> Going forward the database description will be stored in a note with custom type "Database Description"

Thanks for the hint, I knew you mentioned it I don't expect users will want to search to hard to find out how so maybe hard-code a default message that only shows if the custom note type does not exist that states:

**To show a Family tree description here add a note with a custom type of "Database Description"**

**Comment by PQYPLZXHGF** [comment]
The default size of Gramps when started initially looks like this:

![](https://user-images.githubusercontent.com/212355/174416953-be0571ba-07d2-4ea1-990d-ee3f0d107fe0.png)

A little difficult to read, can the description and owner text flow under the first block of text?

**Comment by cdhorn** [comment]
> so maybe hard-code a default message that only shows if the custom note type does not exist that states

Oh if there is no note and you right click for the context menu it will open the note editor with a new note of custom type "Database Description" which I suppose has the constraint that it needs to be in English.  I could probably put something in the [] though telling the user to do that.

**Comment by emyoulation** [comment]
I don't know what policy is for the sidebar & bottombar in the dashboard view. They are entirely disabled for the default Dashboard view mode.

Maybe they should be disabled for the Dashboard CardView mode too?

**Comment by PQYPLZXHGF** [comment]
> Oh if there is no note and you right click for the context menu it will open the note editor with a new note of custom type "Database Description" 

Lol some indication would be appreciated , I didn't consider/find the hidden context menu as I don't see a context menu anywhere else on the dashboard card view, 

Maybe a big yellow (lol) button that shows the words "Add Description" 😁

My new favorite words are "Mystery meat navigation"!

**Comment by cdhorn** [comment]
> can the description and owner text flow under the first block of text?

That would leave a vast tundra of empty real estate on my smaller monitor.

> Maybe they should be disabled for the Dashboard CardView mode too?

Yes, I agree it would be best. I need to look at the default dashboard and see how to go about that.

**Comment by cdhorn** [comment]
> I didn't consider/find the hidden context menu as I don't see a context menu anywhere else on the dashboard card view,

The ctrl-right and ctrl-click contexts are active, but yeah didn't have any ideas for the normal one yet.

**Comment by PQYPLZXHGF** [comment]
> That would leave a vast tundra of empty real estate on my smaller monitor.

>  sidebar & bottombar in the dashboard view. They are entirely disabled for the default Dashboard view mode.

Disabled   sidebar & bottombars would work better 👍  or maybe a line seperator "**|**"  or color difference for each block of text in the family tree summary section or both?  The Dashboard  card is the least colorful out of the cardviews. 

 >The ctrl-right and ctrl-left click contexts are active, but yeah didn't have any ideas for the normal one yet.

Reading the other comment I see at least one suggestions have been given 

* Hide all sections [nodes?]
* Show all sections [nodes?]
* Add/change family tree description.

**Comment by cdhorn** [comment]
> Going forward the database description will be stored in a note with custom type "Database Description"

I have changed this custom note type to "Tree Description" as that seems more apt if someday we have multiple trees in a database.

> is there an Event Roles stat panel?

Second iteration done, some new panels introduced and many different roles and type counts now exposed as well.

**Comment by cdhorn** [comment]
And yes for those who notice, some of the % calcs are not really all that accurate, that will get sorted out soon.

**Comment by cdhorn** [comment]
Percentage calculations *should* all be accurate now.

**Comment by daleathan** [comment]
@cdhorn If you have time, you might be interested Ben Crowder's idea of using [sparklines](https://en.wikipedia.org/wiki/Sparkline) to show information overviews?  Do you think something like this would be suitable for any of the Card views?

Genealogy sparklines
Monday, August 14, 2006
https://bencrowder.net/blog/2006/genealogy-sparklines/


Sparkline pedigree chart
Wednesday, July 23, 2014
https://bencrowder.net/blog/2014/sparkline-pedigree-chart/

Lately I’ve been playing around with implementing genealogy sparklines on the web.
Friday, August 20, 2021
https://bencrowder.net/blog/2021/1259/

Family sparklines (Monday, June 20, 2022)
https://bencrowder.net/blog/2022/family-sparklines/


Found recently via a [Gramps feature request](https://gramps-project.org/bugs/view.php?id=5561)

**Comment by cdhorn** [comment]
@daleathan thanks for heads up on that, always good to see new ideas. I don't have time to consider it at the present moment but will keep it in mind, and of course would accept a PR if someone else has time to do so.

**Comment by cdhorn** [comment]
The last changed gramplet has been converted into a dashboard view to complement the statistics one.  Closing this as I think having these two suffices for now.

**Comment by PQYPLZXHGF** [comment]
> The last changed gramplet has been converted into a dashboard view to complement the statistics one.

I like that it is a view, the only things I'll note are:
* that now it is only a view and I'm no longer able to detach it like a gramplet and have it over to side of the screen :(
* on the dashboard the icons are identical for each view
* on the dashboard the the first icon is now the changes view instead of the statistics view!
* on the changes for repositories I'm seeing alot of spaces instead of a single line per result?

Maybe consider including both the gramplet version and the dashboard view?
